module/builder/collection/builder.go

```diff
@@ -126,66 +133,21 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		// keep track of transactions to enforce rate limiting
 		limiter := newRateLimiter(b.config, parent.Height+1)
 
-		// look up previously included transactions in UN-FINALIZED ancestors
-		ancestorID := parentID
-		clusterFinalID := clusterFinal.ID()
-		for ancestorID != clusterFinalID {
-			ancestor, err := b.clusterHeaders.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor header (%x): %w", ancestorID, err)
-			}
-
-			if ancestor.Height <= clusterFinal.Height {
-				return fmt.Errorf("should always build on last finalized block")
-			}
-
-			payload, err := b.payloads.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor payload (%x): %w", ancestorID, err)
-			}
-
-			collection := payload.Collection
-			for _, tx := range collection.Transactions {
-				lookup.addUnfinalizedAncestor(tx.ID())
-				limiter.addAncestor(ancestor.Height, tx)
-			}
-			ancestorID = ancestor.ParentID
+		// first, look up previously included transactions in UN-FINALIZED ancestors
+		err = b.populateUnfinalizedAncestryLookup(parentID, clusterFinal.Height, lookup, limiter)
```

Comment:
Do we have a cap on the number of finalized cluster blocks to traverse back regardless finalized or unfinalized?

In the event of finalization slow or finalization halt, and cluster chain grows at the normal speed, then it will make the range of finalized cluster blocks very big. In this case, we might look up too many cluster blocks for checking dups. In this worst case, are checking so many cluster blocks indeed necessary to prevent dups?